### PR TITLE
MLCA

### PR DESCRIPTION
**Pull request by beyeler**
Implementation of MLCA of Brero et al.

Some tests are missing, but the implementation is complete.

### CODE REVIEW

**Comment by beyeler** [comment]
You might want to merge the CCG PR before you start looking at this PR because it is also based on the CCG branch.

**Comment by beyeler** [comment]
Tests added. Note that the configuration of the kernels is heavily depending on the domain. Therfore I do not provide any default values for this. For GSVM/LSVM/MRVM I will add some basic configurations to sats.

**Comment by beyeler** [comment]
I added a matching branch in sats 'feat-integrate-mechlib-interaction-mlca' to maintain compatibility.

File: src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/mlca/phases/ExactMLQueryAuctionRoundBuilder.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+	@Override
+	public AuctionRound<BundleExactValueBids> build() {
+		// TODO check if all interactions completed
+		return new MLQueryAuctionRound<>(this.getAuction(), new BundleExactValueBids(interactions.entrySet().stream().collect(Collectors.toMap(e -> this.getAuction().getBidder(e.getKey()), e -> e.getValue().getBid(),(e1,e2)->e1, LinkedHashMap::new))),marginalsToQueryNext,seedNextRound);
+	}
+
+	@Override
+	protected Outcome computeTemporaryResult(OutcomeRuleGenerator outcomeRuleGenerator) {
+		// TODO Auto-generated method stub
```
Comment (islerfab) [inline src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/mlca/phases/ExactMLQueryAuctionRoundBuilder.java:48]: Commenting such that this doesn't get lost (either properly implement it or throw an error that this is not yet implemented)

File: src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/mlca/phases/ExactRandomQueryAuctionRoundBuilder.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	@Override
+	public AuctionRound<BundleExactValueBids> build() {
+		return new RandomQueryAuctionRound<>(this.getAuction(),
+				new BundleExactValueBids(this.interactions.entrySet().stream().collect(
+						Collectors.toMap(e -> this.getAuction().getBidder(e.getKey()), e -> e.getValue().getBid(),(e1,e2)->e1,LinkedHashMap::new))));
+	}
+
+	@Override
+	protected Outcome computeTemporaryResult(OutcomeRuleGenerator outcomeRuleGenerator) {
+		// TODO Auto-generated method stub
```
Comment (islerfab) [inline src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/mlca/phases/ExactRandomQueryAuctionRoundBuilder.java:41]: Commenting such that this doesn't get lost (either properly implement it or throw an error that this is not yet implemented)

Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/mlca/phases/ExactRandomQueryAuctionRoundBuilder.java:41]: Removed the method completly as it does not make sense in general to compute a temporary outcome without knowledge of past rounds. If temporary results are needed this should be rather added in the Auction itself in my opinion.

File: src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/interactions/Interaction.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import org.marketdesignresearch.mechlib.mechanism.auctions.Auction;
+
+/**
+ * Also check activity rules and validity here for submitted bids
+ * 
+ * @author Manuel Beyeler
+ *
+ * @param <T>
+ */
+public interface Interaction {
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/interactions/Interaction.java:13]: added TODO

File: src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/Auction.java [hunk 4]
```diff
@@ -64,191 +100,128 @@ public Good getGood(UUID id) {
...
+        } else {
+        	this.currentPhaseNumber = 0;
+        	this.currentPhaseRoundNumber = 0;
+        }
+        this.prepareNextAuctionRoundBuilder();
     }
 
     /**
      * By default, the bids for the auction results are aggregated in each round
+     * TODO: does it make sense to store outcomes in the auction rounds? rounds can be accessed directly and then some rounds contain outcomes some other don't
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/Auction.java:213]: Removed TODO. Lazy calculation would be for sure a good idea, but I'm not sure if the results should be stored in the AuctionRound itself.

File: src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/Auction.java [hunk 5]
```diff
@@ -1,55 +1,91 @@
...
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
 
-@ToString(callSuper = true) @EqualsAndHashCode(callSuper = true)
+@ToString(callSuper = true) 
+@EqualsAndHashCode(callSuper = true)
 @Slf4j
-public class Auction extends Mechanism implements AuctionInstrumentationable {
+public abstract class Auction<BB extends BundleValueBids<?>> extends Mechanism implements AuctionInstrumentationable {
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/Auction.java:32]: added TODO

File: src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/Auction.java [hunk 6]
```diff
@@ -64,191 +100,128 @@ public Good getGood(UUID id) {
...
+    /**
+     * Reset the auction such that it is in the state of executing the given round. 
+     * (I.e. {@link #getNumberOfRounds()} will return round-1 after this method has executed.
+     * 
+     * @param round the round that should be executed next by this auction
+     */
+    public void resetToRound(int round) {
+        Preconditions.checkArgument(round >= 1 && round <= rounds.size()+1);
+        rounds = rounds.subList(0, round-1);
+        if(round > 0) {
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/Auction.java:201]: You are right removed else branch

File: src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/Auction.java [hunk 7]
```diff
@@ -1,55 +1,91 @@
...
-    public Auction(Domain domain, OutcomeRuleGenerator outcomeRuleGenerator) {
+    public Auction(Domain domain, OutcomeRuleGenerator outcomeRuleGenerator, AuctionPhase<BB> firstPhase) {
         super();
         this.domain = domain;
         this.outcomeRuleGenerator = outcomeRuleGenerator;
-        current = new AuctionRoundBuilder(outcomeRuleGenerator);
-        current.setMipInstrumentation(getMipInstrumentation());
+        this.phases.add(firstPhase);
+        this.prepareNextAuctionRoundBuilder();
+        // TODO
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/Auction.java:66]: done

File: src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/mlca/svr/DistributedSVR.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	@Setter
+	private MipInstrumentation mipInstrumentation;
+
+	
+	public DistributedSVR(Kernel kernel) {
+		this(new SupportVectorSetup(kernel));
+	}
+	
+	
+	/*
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/mlca/svr/DistributedSVR.java:30]: removed

File: src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/mlca/MachineLearningComponent.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package org.marketdesignresearch.mechlib.mechanism.auctions.mlca;
+
+import org.marketdesignresearch.mechlib.core.bid.bundle.BundleValueBids;
+
+public interface MachineLearningComponent<T extends BundleValueBids<?>> {
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/mlca/MachineLearningComponent.java:5]: added TODO

**Comment by beyeler** [comment]
Closing this branch without merge as it contains some conflicts. Upstream branches contains all relevant changes.

**Pull request merged**